canonical/prometheus-k8s-operator
# add new public method update_scrape_job_spec

## Issue


Our current approach in "MetricEndpointProvider" of building everything with the object constructor is not very handy when the charm that instruments the `prometheus_scrape` library is related to an ingress (Traefik for example). 

Let's say we have this scenario:


```mermaid
flowchart LR
    A[prometheus] ---|metrics-endpoint| B[loki]
    B --- |ingress-per-unit| C[traefik]
    
```

And the following code in the loki charm.py:

```python
    def __init__(self, *args):
        ...
        external_url = urlparse(self._external_url)

        scrape_jobs = [
            {
                "metrics_path": f"{external_url.path}/metrics",
                "static_configs": [{"targets": [f"{external_url.hostname}:{external_url.port}"]}],
            }
        ]
        self.metrics_provider = MetricsEndpointProvider(
            self,
            jobs=self.scrape_jobs,
            refresh_event=[
                self.on.update_status,
                self.ingress_per_unit.on.ready_for_unit,
                self.ingress_per_unit.on.revoked_for_unit,
                self.on.ingress_relation_departed,
            ],
        )

...

    @property
    def _external_url(self) -> str:
        """Return the external hostname to be passed to ingress via the relation."""
        if ingress_url := self.ingress_per_unit.url:
            logger.debug("This unit's ingress URL: %s", ingress_url)
            return ingress_url

        return f"http://{self.hostname}:{self._port}"
```

When the relations between the 3 charms are established, prometheus will have loki's scrape job with the external URL and path traefik provides to loki.

```yaml
- honor_labels: true
  job_name: juju_cos_9ba3b5f2_loki_prometheus_scrape
  metrics_path: /cos-loki-0/metrics
  relabel_configs:
  - regex: (.*)
    separator: _
    source_labels:
    - juju_model
    - juju_model_uuid
    - juju_application
    target_label: instance
  static_configs:
  - labels:
      juju_application: loki
      juju_charm: loki-k8s
      juju_model: cos
      juju_model_uuid: 9ba3b5f2-b859-4544-8455-503ec5e5b815
    targets:
    - 192.168.122.10:80
```


But if we remove the relation between [loki](https://github.com/canonical/loki-k8s-operator) and [traefik](https://github.com/canonical/traefik-k8s-operator), prometheus will not receive the new loki URL (and internal IP or FDQN) since at the moment we instantiate `MetricsEndpointProvider` in `__init__` (when `self.ingress_per_unit.on.revoked_for_unit` event is fired) the `self.ingress_per_unit.url` still returns an external url.    

## Solution


We implemented a public method that let us update scrape jobs.
We are following the same pattern we have in [`LokiPushApiProvider`'s `update_endpoint`](https://github.com/canonical/loki-k8s-operator/blob/main/lib/charms/loki_k8s/v0/loki_push_api.py#L1232) and in [`Grafana`'s `update_source`](https://github.com/canonical/grafana-k8s-operator/blob/main/lib/charms/grafana_k8s/v0/grafana_source.py#L412).

With this implementation, we may have in the charm that uses `MetricsEndpointProvider`:

```python
    def __init__(self, *args):
        ...
        self.framework.observe(self.ingress_per_unit.on.ready_for_unit, self._on_ingress_changed)
        self.framework.observe(self.ingress_per_unit.on.revoked_for_unit, self._on_ingress_changed)

        ...

        self.metrics_provider = MetricsEndpointProvider(
            self,
            jobs=self.scrape_jobs,
            refresh_event=[
                self.on.update_status,
                self.ingress_per_unit.on.ready_for_unit,
                self.ingress_per_unit.on.revoked_for_unit,
                self.on.ingress_relation_departed,
            ],
        )

    def _on_ingress_changed(self, _):
        ...
        self.metrics_provider.update_scrape_job_spec(self.scrape_jobs)

    @property
    def scrape_jobs(self) -> list:
        """Generate scrape jobs."""
        if not self.ingress_per_unit.url:
            return [
                {
                    "static_configs": [{"targets": [f"*:{self._port}"]}],
                }
            ]

        external_url = urlparse(self.ingress_per_unit.url)

        return [
            {
                "metrics_path": f"{external_url.path}/metrics",
                "static_configs": [{"targets": [f"{external_url.hostname}:{external_url.port}"]}],
            }
        ]
```
 

## Context


See [this Loki issue](https://github.com/canonical/loki-k8s-operator/issues/210)

## Testing Instructions


- Deploy prometheus, loki and traefik
- Relate prometheus and loki
  - Check that loki scrape job in prometheus config use internal IP or a FQDN
- Relate loki and traefik
  - Check that loki scrape job in prometheus config use external URL provided by traefik.
- Remove the relation between loki and traefik
  - Check that loki scrape job in prometheus config use internal IP or a FQDN again
